Dispatcher allows registering with public = bind, but only the first time

When registering with the dispatcher with a public address equal to the bind address, a duplicate registration takes place (same address, same port). This only happens on the first registration, and future ones will return an error when attempting to do the registration.

Client code to reproduce:
```
package main

import (
        "fmt"
        "os"

        "github.com/scionproto/scion/go/lib/addr"
        "github.com/scionproto/scion/go/lib/snet"
        "github.com/scionproto/scion/go/lib/xtest"
)

var ia = xtest.MustParseIA("1-ff00:0:1")
var addrA, _ = snet.AddrFromString("1-ff00:0:1,[127.0.0.1]:10001")
var addrB, _ = snet.AddrFromString("1-ff00:0:1,[127.0.0.1]:10002")

func main() {
        var err error
        err = snet.Init(ia, "", "")
        if err != nil {
                fmt.Println(err)
                os.Exit(1)
        }
        _, err = snet.ListenSCIONWithBindSVC("udp4", addrA, addrA, addr.SvcPS)
        fmt.Println("A", err)
        _, err = snet.ListenSCIONWithBindSVC("udp4", addrB, addrB, addr.SvcPS)
        fmt.Println("B", err)
}
```

This causes the following log:
```
2018-08-10 14:26:54.006084+0000 [INFO] (7970:dispatcher:dispatcher.c:557) UDP registration request
2018-08-10 14:26:54.006090+0000 [INFO] (7970:dispatcher:dispatcher.c:590) registration for isd_as 1-280375465082881
2018-08-10 14:26:54.006098+0000 [INFO] (7970:dispatcher:dispatcher.c:617) registration for 127.0.0.1:10001
2018-08-10 14:26:54.006104+0000 [INFO] (7970:dispatcher:dispatcher.c:620) bind addr 127.0.0.1:10001
2018-08-10 14:26:54.006110+0000 [INFO] (7970:dispatcher:dispatcher.c:634) SCMP registration included
2018-08-10 14:26:54.006115+0000 [INFO] (7970:dispatcher:dispatcher.c:646) SVC (1) registration included
2018-08-10 14:26:54.006122+0000 [INFO] (7970:dispatcher:dispatcher.c:672) Adding Bind SVC entry. SVC:1 ISD_AS:1-280375465082881 Host:127.0.0.1
2018-08-10 14:26:54.006128+0000 [INFO] (7970:dispatcher:dispatcher.c:676) Adding SVC entry. SVC:1 ISD_AS:1-280375465082881 Host:127.0.0.1
2018-08-10 14:26:54.006135+0000 [INFO] (7970:dispatcher:dispatcher.c:726) assigned port 10001
2018-08-10 14:26:54.006143+0000 [INFO] (7970:dispatcher:dispatcher.c:726) assigned port 10001
2018-08-10 14:26:54.006248+0000 [INFO] (7970:dispatcher:dispatcher.c:497) new socket created: 13
2018-08-10 14:26:54.006266+0000 [INFO] (7970:dispatcher:dispatcher.c:539) received registration for proto: 17 (26 bytes)
2018-08-10 14:26:54.006273+0000 [INFO] (7970:dispatcher:dispatcher.c:557) UDP registration request
2018-08-10 14:26:54.006278+0000 [INFO] (7970:dispatcher:dispatcher.c:590) registration for isd_as 1-280375465082881
2018-08-10 14:26:54.006285+0000 [INFO] (7970:dispatcher:dispatcher.c:617) registration for 127.0.0.1:10002
2018-08-10 14:26:54.006291+0000 [INFO] (7970:dispatcher:dispatcher.c:620) bind addr 127.0.0.1:10002
2018-08-10 14:26:54.006298+0000 [INFO] (7970:dispatcher:dispatcher.c:634) SCMP registration included
2018-08-10 14:26:54.006303+0000 [INFO] (7970:dispatcher:dispatcher.c:646) SVC (1) registration included
2018-08-10 14:26:54.006310+0000 [INFO] (7970:dispatcher:dispatcher.c:726) assigned port 10002
2018-08-10 14:26:54.006316+0000 [ERROR] (7970:dispatcher:dispatcher.c:716) requested port number 10002 not available
2018-08-10 14:26:54.006345+0000 [ERROR] (7970:dispatcher:dispatcher.c:1163) error on socket 13: Invalid argument
```

--- a/c/dispatcher/dispatcher.c
+++ b/c/dispatcher/dispatcher.c
@@ -136,7 +136,8 @@ void handle_app();
 void register_udp(uint8_t *buf, int len, int sock);
 Entry * parse_request(uint8_t *buf, int len, int proto, int sock);
 int add_bind_addr(Entry *e, uint8_t *buf, isdas_t isd_as, int offset);
-int find_available_port(Entry *list, L4Key *key);
+int find_available_udp_port(L4Key *key);
+int find_available_bind_port(L4Key *key);
 void reply(int sock, int port);
 static inline uint16_t get_next_port();
 
@@ -556,9 +557,15 @@ void register_udp(uint8_t *buf, int len, int sock)
 {
     zlog_info(zc, "UDP registration request");
     Entry *e = parse_request(buf, len, L4_UDP, sock);
+    if (memcmp(&e->l4_key, &e->bind_key, sizeof(L4Key)) == 0) {
+        zlog_error(zc, "Not supported same public and bind address");
+        reply(sock, 0);
+        cleanup_socket(sock, num_sockets - 1, EINVAL);
+        return;
+    }
     if (!e)
         return;
-    if (find_available_port(udp_port_list, &e->l4_key) < 0) {
+    if (find_available_udp_port(&e->l4_key) < 0) {
         reply(sock, 0);
         cleanup_socket(sock, num_sockets - 1, EINVAL);
         return;
@@ -568,7 +575,7 @@ void register_udp(uint8_t *buf, int len, int sock)
 
     /* Register bind address info if the app has a bind address */
     if (IS_BIND_SOCKET(*buf)) {
-        if (find_available_port(bind_udp_port_list, &e->bind_key) < 0) {
+        if (find_available_bind_port(&e->bind_key) < 0) {
             reply(sock, 0);
             cleanup_socket(sock, num_sockets - 1, EINVAL);
             return;
@@ -698,7 +705,7 @@ int add_bind_addr(Entry *e, uint8_t *buf, isdas_t isd_as, int offset)
     return offset + port_len + type_len + b_addr_len;
 }
 
-int find_available_port(Entry *list, L4Key *key)
+int find_available_port(bool bind, L4Key *key)
 {
     Entry *old;
     int requested = 1;
@@ -709,7 +716,11 @@ int find_available_port(Entry *list, L4Key *key)
     int start_port = key->port;
     while (1) {
         // Find an available port number between 1025 and 65535.
-        HASH_FIND(hh, list, key, sizeof(L4Key), old);
+        if (bind) {
+            HASH_FIND(bindhh, bind_udp_port_list, key, sizeof(L4Key), old);
+        } else {
+            HASH_FIND(hh, udp_port_list, key, sizeof(L4Key), old);
+        }
         if (old) {
             if (requested) {
                 // If app requested unavailable port number, reply with failure message
@@ -728,6 +739,16 @@ int find_available_port(Entry *list, L4Key *key)
         }
     }
     return 0;
+}
+
+inline int find_available_udp_port(L4Key *key)
+{
+    return find_available_port(false, key);
+}
+
+inline int find_available_bind_port(L4Key *key)
+{
+    return find_available_port(true, key);
 }
 
 void reply(int sock, int port)

